app/controllers/uploads_controller.rb

```diff
@@ -97,32 +103,75 @@ def klass
   end
 
   def requirements_messages
-    # this is a call to custom validators that are not listed inside the class
-    custom_batch_validator_messages = "#{klass.name}Validator::REQUIREMENT_DESCRIPTIONS".safe_constantize
     [Upload.valid_col_seps]
-      .push(*validation_messages)
-      .push(*custom_batch_validator_messages)
-      .compact
+      .push(*validation_messages_presence)
+      .push(*validation_messages_numericality)
+      .push(*validation_messages_uniqueness)
+  end
+
+  def batch
+    custom_batch_validator_messages = {
+      message: 'Requirement Description:',
+      value: "#{klass.name}Validator::REQUIREMENT_DESCRIPTIONS".safe_constantize
+    }
+    if custom_batch_validator_messages[:value].nil?
+      {}
+    else
+      [custom_batch_validator_messages]
+    end
   end
 
-  def validation_messages
+  def klass_validator(validation_class)
+    klass.validators.map do |validations|
+      generic_requirement_message(validations) if validation_class == validations.class
+    end.compact
+  end
+
+  def validation_messages_presence
+    presence = { message: 'These columns must have a value: ', value: [] }
+
+    presence[:value] = klass_validator(ActiveRecord::Validations::PresenceValidator)
+
+    return [presence] unless presence[:value].empty?
+  end
+
+  def validation_messages_numericality
+    numericality = { message: 'These columns can only contain numeric values: ', value: [] }
+
+    numericality[:value] = klass_validator(ActiveModel::Validations::NumericalityValidator)
+
+    return [numericality] unless numericality[:value].empty?
+  end
+
+  def validation_messages_uniqueness
+    uniqueness = { message: 'These columns should contain unique values: ', value: [] }
+
+    uniqueness[:value] = klass_validator(ActiveRecord::Validations::UniquenessValidator)
+
+    return [uniqueness] unless uniqueness[:value].empty?
+  end
+
+  def validation_messages_inclusion
+    inclusion = []
+
     klass.validators.map do |validations|
       case validations
-      when ActiveRecord::Validations::PresenceValidator
-        generic_requirement_message('These columns must have a value: ', validations)
       when ActiveModel::Validations::InclusionValidator
-        inclusion_requirement_message(validations)
-      when ActiveModel::Validations::NumericalityValidator
-        generic_requirement_message('These columns can only contain numeric values: ', validations)
-      when ActiveRecord::Validations::UniquenessValidator
-        generic_requirement_message('These columns should contain unique values: ', validations)
+        array = { message: affected_attributes(validations), value: inclusion_requirement_message(validations) }
+        inclusion.push(array)
       end
     end
+
+    if !inclusion.empty?
+      [inclusion]
+    else
+      inclusion = {}
+    end
```

Comment:
Change the if/else block to

```suggestion
    [inclusion] unless inclusion.empty?
```